File: isp-billing-system-BACKEND/src/ai/mlrModel.js
/**
 * mlrModel.js – Multiple Linear Regression (OLS) for Revenue Prediction
 *
 * Features (X): activeSubscribers, avgDataUsageMB, paymentDelays, planScore
 * Target  (y): totalRevenue (KES)
 *
 * Method: Normal Equation  β = (XᵀX)⁻¹ Xᵀy
 * Implemented with plain JavaScript – no external ML libraries.
 */

'use strict';

/* ──────────────────────────────────────────────────────────────
   Lightweight matrix helpers
   ────────────────────────────────────────────────────────────── */

/** Multiply two matrices A (m×k) × B (k×n) → m×n */
function matMul(A, B) {
  const m = A.length, k = A[0].length, n = B[0].length;
  const C = Array.from({ length: m }, () => new Array(n).fill(0));
  for (let i = 0; i < m; i++)
    for (let j = 0; j < n; j++)
      for (let l = 0; l < k; l++)
        C[i][j] += A[i][l] * B[l][j];
  return C;
}

/** Transpose matrix */
function matT(A) {
  return A[0].map((_, j) => A.map(row => row[j]));
}

/** Gauss-Jordan inversion of a square matrix */
function matInv(A) {
  const n = A.length;
  // Build augmented [A | I]
  const M = A.map((row, i) => {
    const aug = [...row.map(v => v), ...new Array(n).fill(0)];
    aug[n + i] = 1;
    return aug;
  });

  for (let col = 0; col < n; col++) {
    // Find pivot
    let maxRow = col;
    for (let r = col + 1; r < n; r++)
      if (Math.abs(M[r][col]) > Math.abs(M[maxRow][col])) maxRow = r;
    [M[col], M[maxRow]] = [M[maxRow], M[col]];

    const pivot = M[col][col];
    if (Math.abs(pivot) < 1e-12) throw new Error('Matrix is singular – cannot invert');
    for (let j = 0; j < 2 * n; j++) M[col][j] /= pivot;

    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const factor = M[r][col];
      for (let j = 0; j < 2 * n; j++) M[r][j] -= factor * M[col][j];
    }
  }

  return M.map(row => row.slice(n));
}

/* ──────────────────────────────────────────────────────────────
   MLR Model state  (persisted in memory between calls)
   ────────────────────────────────────────────────────────────── */

let coefficients = null; // [β0, β1, β2, β3, β4]
let rSquared = 0;
let residualStdErr = 0;
let trainingSamples = 0;
let lastTrained = null;

/* ──────────────────────────────────────────────────────────────
   Feature engineering helpers
   ────────────────────────────────────────────────────────────── */

/**
 * Converts plan distribution map → weighted quality score.
 * Premium plans are weighted higher.
 */
function computePlanScore(planDistribution = {}) {
  const weights = { basic: 1, standard: 2, premium: 3, enterprise: 4 };
  let total = 0, weighted = 0;
  for (const [plan, count] of Object.entries(planDistribution)) {
    const w = weights[plan.toLowerCase()] || 1;
    weighted += w * count;
    total += count;
  }
  return total === 0 ? 1 : weighted / total;
}

/**
 * Build design matrix row: [1, x1, x2, x3, x4]  (bias term included)
 */
function buildRow(activeSubscribers, avgDataUsageMB, paymentDelays, planScore) {
  return [1, activeSubscribers, avgDataUsageMB, paymentDelays, planScore];
}

/* ──────────────────────────────────────────────────────────────
   Training
   ────────────────────────────────────────────────────────────── */

/**
 * Train MLR on historical data rows.
 * @param {Array<{activeSubscribers, avgDataUsageMB, paymentDelays, planDistribution, revenue}>} samples
 * @returns {{ coefficients, rSquared, residualStdErr, n }}
 */
function train(samples) {
  if (!samples || samples.length < 4) {
    throw new Error('MLR requires at least 4 training samples');
  }

  const X = samples.map(s =>
    buildRow(s.activeSubscribers, s.avgDataUsageMB, s.paymentDelays,
      computePlanScore(s.planDistribution))
  );
  const y = samples.map(s => [s.revenue]);

  // β = (XᵀX)⁻¹ Xᵀy
  const Xt = matT(X);
  const XtX = matMul(Xt, X);
  const XtXinv = matInv(XtX);
  const Xty = matMul(Xt, y);
  const beta = matMul(XtXinv, Xty).map(r => r[0]);

  // Predictions & residuals
  const yHat = X.map(row => row.reduce((s, xi, i) => s + xi * beta[i], 0));
  const residuals = y.map((r, i) => r[0] - yHat[i]);

  // R²
  const yMean = y.reduce((s, r) => s + r[0], 0) / y.length;
  const ssTot = y.reduce((s, r) => s + (r[0] - yMean) ** 2, 0);
  const ssRes = residuals.reduce((s, r) => s + r ** 2, 0);
  const r2 = ssTot === 0 ? 1 : 1 - ssRes / ssTot;

  // Residual standard error
  const df = samples.length - beta.length; // n - p
  const rse = Math.sqrt(ssRes / Math.max(df, 1));

  // Persist in module state
  coefficients = beta;
  rSquared = r2;
  residualStdErr = rse;
  trainingSamples = samples.length;
  lastTrained = new Date().toISOString();

  return { coefficients, rSquared, residualStdErr, n: trainingSamples };
}

/* ──────────────────────────────────────────────────────────────
   Prediction
   ────────────────────────────────────────────────────────────── */

/**
 * Predict next month's revenue.
 * Returns predicted value + 95 % confidence interval + feature importance.
 */
function predict(activeSubscribers, avgDataUsageMB, paymentDelays, planDistribution = {}) {
  if (!coefficients) {
    throw new Error('Model has not been trained yet. Call /api/ai/retrain first.');
  }

  const planScore = computePlanScore(planDistribution);
  const row = buildRow(activeSubscribers, avgDataUsageMB, paymentDelays, planScore);
  const predicted = row.reduce((s, xi, i) => s + xi * coefficients[i], 0);

  // 95 % CI  ≈  ŷ ± 1.96 × RSE  (large-sample approximation)
  const margin = 1.96 * residualStdErr;
  const confidenceLow = Math.max(0, predicted - margin);
  const confidenceHigh = predicted + margin;

  // Feature importance (absolute contribution relative to total)
  const featureNames = ['bias', 'activeSubscribers', 'avgDataUsageMB', 'paymentDelays', 'planScore'];
  const contributions = row.map((xi, i) => Math.abs(xi * coefficients[i]));
  const totalContrib = contributions.reduce((a, b) => a + b, 0);

  const influencingFactors = featureNames
    .map((name, i) => ({
      factor: name,
      coefficient: parseFloat(coefficients[i].toFixed(4)),
      contribution: totalContrib > 0
        ? parseFloat(((contributions[i] / totalContrib) * 100).toFixed(2))
        : 0
    }))
    .filter(f => f.factor !== 'bias')
    .sort((a, b) => b.contribution - a.contribution);

  return {
    predictedRevenue: parseFloat(predicted.toFixed(2)),
    confidenceInterval: {
      low: parseFloat(confidenceLow.toFixed(2)),
      high: parseFloat(confidenceHigh.toFixed(2)),
      level: '95%'
    },
    influencingFactors,
    modelStats: {
      rSquared: parseFloat(rSquared.toFixed(4)),
      residualStdErr: parseFloat(residualStdErr.toFixed(2)),
      trainingSamples,
      lastTrained
    },
    inputs: { activeSubscribers, avgDataUsageMB, paymentDelays, planScore }
  };
}

/* ──────────────────────────────────────────────────────────────
   Default coefficients (fallback before first training)
   Derived from domain knowledge for a Kenyan ISP
   β0=intercept, β1=subscribers, β2=data, β3=delays, β4=planScore
   ────────────────────────────────────────────────────────────── */
function initDefaultCoefficients() {
  // KES: base 50,000 + 800/subscriber + 0.5/MB + -200/delay + 3000/planScore
  coefficients = [50000, 800, 0.5, -200, 3000];
  rSquared = 0;
  residualStdErr = 15000;
  trainingSamples = 0;
  lastTrained = 'default (not trained)';
}

initDefaultCoefficients();

/* ──────────────────────────────────────────────────────────────
   Exports
   ────────────────────────────────────────────────────────────── */
module.exports = {
  train,
  predict,
  getModelState: () => ({ coefficients, rSquared, residualStdErr, trainingSamples, lastTrained }),
  computePlanScore
};

File: ai-service/models/mlr_model.py
"""
mlr_model.py – Multiple Linear Regression for Revenue Prediction
            using NumPy only (no scikit-learn).

Method: Ordinary Least Squares via the Normal Equation
        β = (XᵀX)⁻¹ Xᵀy

Features
--------
x1 : active_subscribers     – number of paying subscribers that month
x2 : avg_data_usage_mb      – mean data consumed per subscriber (MB)
x3 : payment_delays         – count of late/missed payments
x4 : plan_score             – weighted quality index from plan distribution

Target
------
y  : revenue                – total collected revenue (KES)
"""

import logging
from datetime import datetime
from typing import Optional

import numpy as np

from models.model_store import load_model, save_model

logger = logging.getLogger(__name__)

# ── Model name for persistence ───────────────────────────────────────────────
_MODEL_KEY = "mlr"

# ── Default heuristic coefficients (KES ISP domain knowledge) ───────────────
#   [β0,   β1,   β2,  β3,    β4  ]
#   [base, /sub, /MB, /delay, /plan_score]
_DEFAULT_COEF = np.array([50_000.0, 800.0, 0.5, -200.0, 3_000.0])
_DEFAULT_RSE = 15_000.0

# ── Module state ─────────────────────────────────────────────────────────────
_coef: np.ndarray = _DEFAULT_COEF.copy()
_r_squared: float = 0.0
_residual_std_err: float = _DEFAULT_RSE
_training_samples: int = 0
_last_trained: str = "default (not trained)"


# ── Feature helpers ──────────────────────────────────────────────────────────

def _plan_score(plan_distribution: dict) -> float:
    """
    Convert plan distribution counts → weighted quality score.
    Premium plans carry higher weight.
    """
    weights = {"basic": 1, "standard": 2, "premium": 3, "enterprise": 4}
    total = 0
    weighted = 0.0
    for plan, count in (plan_distribution or {}).items():
        w = weights.get(plan.lower(), 1)
        weighted += w * count
        total += count
    return weighted / total if total > 0 else 1.0


def _build_design_row(
    active_subscribers: float,
    avg_data_usage_mb: float,
    payment_delays: float,
    ps: float,
) -> list[float]:
    """Build one design-matrix row: [1, x1, x2, x3, x4]."""
    return [1.0, active_subscribers, avg_data_usage_mb, payment_delays, ps]


# ── Persistence boot-strap ───────────────────────────────────────────────────

def _boot() -> None:
    """Load saved coefficients from disk on module import."""
    global _coef, _r_squared, _residual_std_err, _training_samples, _last_trained
    saved = load_model(_MODEL_KEY)
    if saved and "coef" in saved:
        _coef = np.array(saved["coef"])
        _r_squared = saved.get("r_squared", 0.0)
        _residual_std_err = saved.get("residual_std_err", _DEFAULT_RSE)
        _training_samples = saved.get("training_samples", 0)
        _last_trained = saved.get("last_trained", "loaded from disk")
        logger.info(f"[mlr] Loaded persisted coefficients (n={_training_samples}).")


_boot()


# ── Training ─────────────────────────────────────────────────────────────────

def train(samples: list[dict]) -> dict:
    """
    Fit OLS regression on historical monthly records.

    Each sample must have:
        active_subscribers, avg_data_usage_mb, payment_delays,
        plan_distribution (dict), revenue (float)

    Requires at least 4 samples (more than the number of parameters).
    """
    global _coef, _r_squared, _residual_std_err, _training_samples, _last_trained

    MIN_SAMPLES = 6
    if len(samples) < MIN_SAMPLES:
        return {
            "status": "skipped",
            "reason": f"Insufficient training data ({len(samples)} samples, need {MIN_SAMPLES})",
            "using": "default_heuristic_coefficients"
        }

    X_rows = []
    y_vals = []
    for s in samples:
        ps = _plan_score(s.get("plan_distribution", {}))
        X_rows.append(_build_design_row(
            float(s["active_subscribers"]),
            float(s["avg_data_usage_mb"]),
            float(s["payment_delays"]),
            ps,
        ))
        y_vals.append(float(s["revenue"]))

    X = np.array(X_rows)          # (n, 5)
    y = np.array(y_vals)          # (n,)

    # Normal equation: β = (XᵀX)⁻¹ Xᵀy
    XtX = X.T @ X
    Xty = X.T @ y
    beta, _, _, _ = np.linalg.lstsq(XtX, Xty, rcond=None)

    y_hat = X @ beta
    residuals = y - y_hat

    y_mean = np.mean(y)
    ss_tot = np.sum((y - y_mean) ** 2)
    ss_res = np.sum(residuals ** 2)
    r2 = 1.0 - ss_res / ss_tot if ss_tot != 0 else 1.0

    df = max(len(samples) - len(beta), 1)
    rse = float(np.sqrt(ss_res / df))

    _coef = beta
    _r_squared = float(r2)
    _residual_std_err = rse
    _training_samples = len(samples)
    _last_trained = datetime.utcnow().isoformat()

    state = {
        "coef": _coef.tolist(),
        "r_squared": _r_squared,
        "residual_std_err": _residual_std_err,
        "training_samples": _training_samples,
        "last_trained": _last_trained,
        "model_version": "1.0.0",
    }
    save_model(_MODEL_KEY, state)

    return {
        "coefficients": _coef.tolist(),
        "r_squared": round(_r_squared, 4),
        "residual_std_err": round(_residual_std_err, 2),
        "n": _training_samples,
    }


# ── Prediction ───────────────────────────────────────────────────────────────

def predict(
    active_subscribers: float,
    avg_data_usage_mb: float,
    payment_delays: float,
    plan_distribution: Optional[dict] = None,
) -> dict:
    """
    Predict next month's revenue.

    Returns
    -------
    dict with keys:
        predicted_revenue, confidence_interval, influencing_factors, model_stats
    """
    ps = _plan_score(plan_distribution or {})
    row = np.array(_build_design_row(active_subscribers, avg_data_usage_mb, payment_delays, ps))

    predicted = float(row @ _coef)

    # 95 % CI  ≈  ŷ ± 1.96 × RSE  (large-sample)
    margin = 1.96 * _residual_std_err
    ci_low = max(0.0, predicted - margin)
    ci_high = predicted + margin

    # Feature importance by absolute contribution
    feature_names = ["bias", "active_subscribers", "avg_data_usage_mb", "payment_delays", "plan_score"]
    contributions = np.abs(row * _coef)
    total_contrib = float(np.sum(contributions)) or 1.0

    influencing_factors = sorted(
        [
            {
                "factor": feature_names[i],
                "coefficient": round(float(_coef[i]), 4),
                "contribution_pct": round(float(contributions[i] / total_contrib * 100), 2),
            }
            for i in range(len(feature_names))
            if feature_names[i] != "bias"
        ],
        key=lambda x: x["contribution_pct"],
        reverse=True,
    )

    return {
        "predicted_revenue": round(predicted, 2),
        "confidence_interval": {
            "low": round(ci_low, 2),
            "high": round(ci_high, 2),
            "level": "95%",
        },
        "influencing_factors": influencing_factors,
        "model_stats": {
            "r_squared": round(_r_squared, 4),
            "residual_std_err": round(_residual_std_err, 2),
            "training_samples": _training_samples,
            "last_trained": _last_trained,
        },
        "inputs": {
            "active_subscribers": active_subscribers,
            "avg_data_usage_mb": avg_data_usage_mb,
            "payment_delays": payment_delays,
            "plan_score": round(ps, 4),
        },
    }


def get_state() -> dict:
    return {
        "coefficients": _coef.tolist(),
        "r_squared": _r_squared,
        "residual_std_err": _residual_std_err,
        "training_samples": _training_samples,
        "last_trained": _last_trained,
    }
